src/contrib/cluster/Akka.Cluster.Sharding/PersistentShardCoordinator.cs

```diff
@@ -153,7 +156,27 @@ public State Updated(IDomainEvent e)
                             if (!Regions.TryGetValue(message.Region, out var shardRegions))
                                 throw new ArgumentException($"Region {message.Region} not registered", nameof(e));
                             if (Shards.ContainsKey(message.Shard))
-                                throw new ArgumentException($"Shard {message.Shard} is already allocated", nameof(e));
+                            {
+                                if (!isRecovering)
```

Comment:
I've decided to implement reader tolerance in only one location for now the `ShardHomeAllocated` message, since this is where the issue has come up most often and it's also probably the safest place to do it.